api/utils/keys/yubikey.go

```diff
@@ -644,150 +728,148 @@ func (y *YubiKey) open() (yk *piv.YubiKey, err error) {
 		// It's also possible that the user is running another PIV program, which may hold the PIV
 		// connection indefinitely (yubikey-agent). In this case, user action is necessary, so we
 		// alert them with this issue.
-		return nil, trace.LimitExceeded("could not connect to YubiKey as another application is using it. Please try again once the program that uses the YubiKey, such as yubikey-agent is closed")
+		return trace.LimitExceeded("could not connect to YubiKey as another application is using it. Please try again once the program that uses the YubiKey, such as yubikey-agent is closed")
 	} else if err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
 	}
-	return yk, nil
+
+	c.waitClose.Add(1)
+	go func() {
+		c.waitClose.Wait()
```

Comment:
Fair point, maybe a waitgroup isn't the best option here, but I haven't thought of something simpler yet.

>  Something like return c.releaseConnection, err so that the caller couldn't avoid it without explicitly ignoring the return value could work

I like that idea

Reply:
I like it too! I also moved `c.releaseConnection` to a function inside `c.connect`. I hope I understood your comment correctly, Nic :)

Reply:
There is actually a race condition here - if the waitgroup is empty and the connection is closed, right as a new connection tries to connect, it would get a nil connection. Since this is my faulty code, here's my suggested fix:

```diff
diff --git a/api/utils/keys/yubikey.go b/api/utils/keys/yubikey.go
index 4a402a311f..2399cdc78e 100644
--- a/api/utils/keys/yubikey.go
+++ b/api/utils/keys/yubikey.go
@@ -336,6 +336,7 @@ func (y *YubiKeyPrivateKey) sign(ctx context.Context, rand io.Reader, digest []b
 		return nil, trace.Wrap(err)
 	}
 	defer release()
+
 	var touchPromptDelayTimer *time.Timer
 	if y.attestation.TouchPolicy != piv.TouchPolicyNever {
 		touchPromptDelayTimer = time.NewTimer(signTouchPromptDelay)
@@ -675,9 +676,9 @@ type sharedPIVConnection struct {
 	card string
 
 	// conn is the shared PIV connection.
-	conn      *piv.YubiKey
-	mu        sync.Mutex
-	waitClose sync.WaitGroup
+	conn              *piv.YubiKey
+	mu                sync.Mutex
+	activeConnections int
 }
 
 // connect a connection to YubiKey PIV module. The returned connection should be closed once
@@ -688,14 +689,30 @@ func (c *sharedPIVConnection) connect() (func(), error) {
 	defer c.mu.Unlock()
 
 	release := func() {
-		go func() {
-			time.Sleep(releaseConnectionDelay)
-			c.waitClose.Done()
-		}()
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.activeConnections--
+
+		// If this is the last active connection, close after a delay,
+		// giving other callers a chance to claim the connection first.
+		if c.activeConnections == 0 {
+			go func() {
+				time.Sleep(releaseConnectionDelay)
+
+				c.mu.Lock()
+				defer c.mu.Unlock()
+
+				// If there are still no new active connections, close the shared connection.
+				if c.activeConnections == 0 {
+					c.conn.Close()
+					c.conn = nil
+				}
+			}()
+		}
 	}
 
 	if c.conn != nil {
-		c.waitClose.Add(1)
+		c.activeConnections++
 		return release, nil
 	}
 
@@ -739,16 +756,7 @@ func (c *sharedPIVConnection) connect() (func(), error) {
 		return nil, trace.Wrap(err)
 	}
 
-	c.waitClose.Add(1)
-	go func() {
-		c.waitClose.Wait()
-
-		c.mu.Lock()
-		defer c.mu.Unlock()
-		c.conn.Close()
-		c.conn = nil
-	}()
-
+	c.activeConnections++
 	return release, nil
 }
```

Reply:
Thanks for the code!
I made minor modifications:
* Before decrementing `activeConnections` I check if the number is greater than 0. There is a chance that someone would call `release` twice.
* In the release goroutine I check if the connection hasn't been closed already. There was a bug in the following scenario:
1. Call `connect` and `release` the first time - the first release goroutine starts (`c.activeConnections == 0`).
2. Call `connect` and `release` the second time - the second release goroutine starts (`c.activeConnections == 0`).
3. The first goroutine closes the connection.
4. The second gouroutine makes an attempt to close the connection too (since `c.activeConnections == 0`) and it panics because it has already been closed.

Reply:
Yeah good catch, shared a different suggestion below to fix the underlying issue.